pip/pip-429.md

```diff
@@ -0,0 +1,161 @@
+
+# Goals
+
+## In Scope
+
+Instead of relying on the `compacted_out` field in `SingleMessageMetadata`, this PIP proposes to use `MessageMetadata` to determine the last message ID in the compacted last entry. Since `compacted_out` is no longer used, the payload buffer's format could be improved as well.
+
+## Out of Scope
+
+# High Level Design
+
+To enhance efficiency and simplify the handling of compacted entries, a new field will be added to `MessageMetadata` to record the batch indexes of all retained single messages. This change allows the server to determine the last message ID directly from the new field in `MessageMetadata` when processing a `GetLastMessageId` request, rather than relying on the `compacted_out` field in `SingleMessageMetadata`.
+
+With this update, the `compacted_out` field will no longer be used, and only the retained messages will be included in the payload buffer. For example, the previous representation:
+
+```yaml
+metadata: # MessageMetadata
+  num_messages_in_batch: 4
+  compression: LZ4
+  compacted_batch_indexes: [1] # Retained messages' batch indexes
+```
+
+The improvements are:
+1. **Reduced Payload Size**: 
+   - In the new format, only the retained messages are included in the payload buffer. 
+   - For the example above, the new format contains just 1 pair of `SingleMessageMetadata` and its corresponding value buffer, compared to 4 pairs in the original format.
+
+2. **Efficient Batch Index Retrieval**:
+   - The actual batch index of retained messages can now be retrieved from the `compacted_batch_indexes` field in `MessageMetadata`. 
+   - In this example, the retained message's batch index is `compacted_batch_indexes[0] = 1`.
+
+## Public-facing Changes
+
+### Binary protocol
+
+A new field will be added to `MessageMetadata`:
+
+```protobuf
+    // Indicates the indexes of messages retained in the batch after compaction. When a batch is compacted, 
+    // some messages may be removed (compacted out). For example, if the original batch contains:
+    // `k0 => v0, k1 => v1, k2 => v2, k1 => null`, the compacted batch will retain only `k0 => v0` and `k2 => v2`.
+    // In this case, this field will be set to `[0, 2]`, and the payload buffer will only include the retained messages.
+    //
+    // Note: Batches compacted by older versions of the compaction service do not include this field. For such batches, 
+    // the `compacted_out` field in `SingleMessageMetadata` must be checked to identify and filter out compacted messages 
+    // (e.g., `k1 => v1` and `k1 => null` in the example above).
+    repeated int32 compacted_batch_indexes = 31;
+```
+
+# Backward & Forward Compatibility
+
```

Comment:
This section is empty. I think we could move the section line 124-126 here.

Reply:
Added